### ISSUE TITLE
Player and question ratings

### ISSUE THREAD

**Issue by rodrigogiraoserrao** [underfitted/bnomial#34]
With regards to the ELO rating of players:

A player has two attributes, namely their `elo` rating and their `k_adjustment_factor`. The `k_adjustment_factor` is a value that is mostly a constant that controls how volatile a player's rating is. Think of the `k_adjustment_factor` as the learning rate, which starts off bigger, so that we get close to the actual rating quicker, but then becomes smaller to prevent wild oscillations. 

To update the ELO of a player, we need to compute two quantities:
 - the expected performance of the player; and
 - the actual performance of the player.

(Both of these are values between 0 and 1.) 

The actual performance of a player with regards to a question is going to be the number of options that the player got right divided by the total number of options that the player could get right.

So, if a question has 4 options and the user got 2 of them right, their actual performance in that question is 0.5.
If a question is a regular single-choice question, then the actual performance is either 0 (the player got it wrong) or 1 (the player got it right).

Now, computing the expected performance of the player is
```py
def expected_perf(player_rating, opponent_rating):
    return 1 / (1 + 10 ** ((opponent_rating - player_rating) / 400))  
```
in Python.

So, if we know the `actual_performance` of the player, and can compute the expected performance, this is how the rating for the player is updated:
```py
player_rating += k_adjustment_factor * (actual_performance - expected_perf(player_rating, question_rating))
```

Updating the player rating after 3 questions is done in pretty much the same way. We compute the actual performance in the three questions, and add those up. We compute the expected performances, and add those up, and then plug them into the formula:
```py
player_rating += k_adjustment_factor * (actual_performance_over_last_questions - expected_performance_over_last_questions)
```

To update the rating of a question, we do the same thing, except it's the other way around. For a question, the actual performance is the number of options that the player got wrong (because the “objective” of a question is for players to answer incorrectly).

For the initial `k_adjustment_factor`, I suggest `64` for the players and `32` for the questions. Then, the values can come down to `16`.
This can happen in plenty of different ways. For players, it can come down by 1 for every question answered. For questions, it can come down by 1 every time the rating of the question is updated.

There is a certain arbitrarity in these values, but the point is that they start bigger and then become smaller.

---

We may want to adjust some of the numerical values mentioned, like the initial/final `k_adjustment_factor` and the “magic constant” `400` in the formula for the expected performance.

The `400` controls the expected difficulty of a question that is `400` rating points away from the player's rating. As is, the formula dictates that for a question that has 4 options and is 400 points higher than the player, the player is only expected to get 1 option correctly. This is an arbitrary threshold, and we might want to move it if we understand that the perceived difficulty doesn't match what the formula computes.

**Comment by svpino** [comment]
Some notes:

* New questions should start with the median score of all other questions.
* Consider taking questions through a beta period so we can calibrate their score before releasing them.

**Comment by svpino** [comment]
Questions we need to answer:

1. What's the initial score of a question? We need to seed the first set of questions with a score.
2. Do we need to retroactively update the score of users?
3. Do we need to retroactively update the score of questions?
4. Before we said that we were going to update the score of a user every 3 answers. Is this still the way we should go?
5. Before we said that we were going to update the score of a question every 10 answers. Is this still the way we should go?

**Comment by RodrigoGiraoSerrao** [comment]
Following up on this, and after [a simulation I ran](https://mathspp.com/blog/elo-rating-system-simulation), I am confident that the ELO-based rating system we devised is going to hold up just fine.
Answering your specific questions @svpino:

 1. Let's go with 1000 initial score for questions and players;
 2. No;
 3. No;
 4. We _can_ do this if we don't want users to know their live rating, but I think it is more rewarding if players can see their ratings change live. So, we either update the score live or show some sort of progress bar saying ”Score will be updated after ? / 3”. However, it is kind of weird to be answering questions, not getting confirmation about your answers being correct/incorrect, and after 3 questions you see your rating arbitrarily going up/down;
 5. Question's rating does _not_ need to be public and, in that sense, I think a question's rating can be updated live. An alternative (based off of what I've seen at chess.com) is to hide the rating of the question until after the player answered it. After answering, we can either show the updating rating and the rating change (which means the user gets to know if they got it right/wrong) or we just show the new question rating, so the user can know they just answered a question of a given rating level.
